core/src/stages/pick.cpp

```diff
@@ -1,63 +1,143 @@
 
 #include <moveit/task_constructor/solvers/cartesian_path.h>
+#include <moveit/task_constructor/solvers/pipeline_planner.h>
 
 #include <moveit/task_constructor/container.h>
 #include <moveit/task_constructor/stages/move_relative.h>
+#include <moveit/task_constructor/stages/grasp_provider.h>
+#include <moveit/task_constructor/stages/place_provider.h>
+#include <moveit/task_constructor/stages/compute_ik.h>
+#include <moveit/task_constructor/stages/modify_planning_scene.h>
+#include <moveit/task_constructor/stages/move_to.h>
 
 #include <moveit/planning_scene/planning_scene.h>
 
+#include <eigen_conversions/eigen_msg.h>
+
 namespace moveit {
 namespace task_constructor {
 namespace stages {
 
-PickPlaceBase::PickPlaceBase(Stage::pointer&& grasp_stage, const std::string& name, bool forward)
-  : SerialContainer(name) {
+PickPlaceBase::PickPlaceBase(const std::string& name, const std::string& provider_stage_plugin_name,  bool is_pick)
+  : SerialContainer(name),
+  grasp_provider_class_loader_("moveit_task_constructor_core", "moveit::task_constructor::stages::GraspProviderBase"),
+  place_provider_class_loader_("moveit_task_constructor_core", "moveit::task_constructor::stages::PlaceProviderBase") {
 	PropertyMap& p = properties();
-	p.declare<std::string>("object", "name of object to grasp");
+	p.declare<std::string>("object", "name of object to grasp/place");
 	p.declare<std::string>("eef", "end effector name");
-	p.declare<std::string>("eef_frame", "name of end effector frame");
+	p.declare<geometry_msgs::PoseStamped>("ik_frame", "frame to be moved towards goal pose");
+	p.declare<std::string>("eef_group_open_pose", "Name of open pose of eef_group");
+	p.declare<std::string>("eef_group_close_pose", "Name of close pose of eef_group");
+	p.declare<std::vector<std::string>>("support_surfaces", {}, "Name of support surfaces");
 
 	// internal properties (cannot be marked as such yet)
 	p.declare<std::string>("eef_group", "JMG of eef");
 	p.declare<std::string>("eef_parent_group", "JMG of eef's parent");
 
 	cartesian_solver_ = std::make_shared<solvers::CartesianPath>();
-	int insertion_position = forward ? -1 : 0;  // insert children at end / front, i.e. normal or reverse order
+	sampling_planner_ = std::make_shared<solvers::PipelinePlanner>();
 
-	auto init_ik_frame = [](const PropertyMap& other) -> boost::any {
-		geometry_msgs::PoseStamped pose;
-		const boost::any& frame = other.get("eef_frame");
-		if (frame.empty())
-			return boost::any();
+	is_pick_ = is_pick;
 
-		pose.header.frame_id = boost::any_cast<std::string>(frame);
-		pose.pose.orientation.w = 1.0;
-		return pose;
-	};
+	provider_stage_plugin_name_ = provider_stage_plugin_name;
 
 	{
-		auto approach = std::make_unique<MoveRelative>(forward ? "approach object" : "retract", cartesian_solver_);
-		PropertyMap& p = approach->properties();
+		auto move_there = std::make_unique<MoveRelative>(is_pick_ ? "approach object" : "lower object", cartesian_solver_);
```

Comment:
`move_there` can be read in an imperative tone, so the intent is not clear upon reading. I'd even call this `approach_stage`, since you also assign it to `approach_stage_` a few lines later.